app/services/tokens_associated_to_email_query.rb

```diff
@@ -0,0 +1,42 @@
+class TokensAssociatedToEmailQuery
+  attr_reader :email, :api
+
+  def initialize(email:, api: nil)
+    @email = email
+    @api = api
+  end
+
+  def call
+    return tokens.valid_for(api).uniq if api
+
+    tokens.uniq
+  end
+
+  private
+
+  def tokens
+    tokens_from_contacts.or(tokens_from_user)
+  end
+
+  def tokens_from_contacts
+    Token
+      .eager_load(:contacts)
+      .where(contacts:)
+  end
+
+  def tokens_from_user
+    Token.where(authorization_request:)
```

Comment:
```suggestion
    Token.joins(:authorization_request, 'authorization_requests.users').where('users.email = (?)', email)
```

pas sûr de la syntaxe

Reply:
L'idée est bonne mais ça rend la requête incompatible avec le `or` ci-dessus, faut d'abord que je vois si on peut faire une seule requête du coup